garage/misc/git_handler.py

```diff
@@ -0,0 +1,336 @@
+            raise
+        return str(commit)
+
+    def create_stash(self):
+        """Stashes the local changes that have not been staged.
+
+        The stash is not stored anywhere in the ref namespace so the user does
+        not have remove it manually later.
+
+        Returns
+        -------
+        The 40 character SHA if local changes were found, or an empty string
+        otherwise.
+
+        """
+        stash_sha = ""
+        try:
+            self.repo.git.diff(exit_code=True)
+        except GitCommandError:
+            # Exit code 1 means there are differences
+            stash_sha = self.repo.git.stash("create")
+            pass
+        return stash_sha
+
+    def get_current_branch_name(self):
+        """Return the name of the checked out branch.
+
+        If the HEAD is in detached state, an empty string is returned.
+
+        Precondition
+        ------------
+        - No branch (local or remote) in the repository is named "HEAD".
+
+        """
+        curr_branch_name = self.repo.git.rev_parse(
+            self.repo.head, abbrev_ref=True)
+        if curr_branch_name == "HEAD":
+            curr_branch_name = ""
+        return curr_branch_name
+
+    def run_experiment_on_target(self, target, exp_args):
+        """Run a experiment on the indicated commit defined by target.
+
+        After the target is verified, if there are any non-staged changes in
+        the working directory, they're stashed, the target commit is checked
+        out, the experiment is executed and the working directory is restored
+        at the end, including the stashed changes.
+
+        Parameters
+        ----------
+        - target: a key-value string, where the key indicates the type of
+            target (tag, branch, sha), and the value the reference for the
+            corresponding type. For example:
+                sha: 8d54s23a
+                branch: my_branch
+                branch: remote/master
+                tag: my_tag
+        - exp_args: it's a dictionary of arguments obtained from the function
+            run_experiment at garage.misc.instrument, that will be passed to
+            the experiment once the target is checked out.
+
+        Precondition
+        ------------
+        - The working directory is not in the middle of a conflict resolution.
+
+        """
+        assert isinstance(target, str), ("The git target has to be a string")
+        self.target = target
+        self.target_sha = self._verify_target(self.target)
+        self.restore_sha = self.get_sha()
+        self.stash_sha = self.create_stash()
+        self.branch_name = self.get_current_branch_name()
+        self.target_branch_name = ("run_exp_on_" + str(int(time.time())))
+        try:
+            self._switch_to_target(self.target_branch_name, self.target_sha)
+            current_sha = self.get_sha()
+            self._run_experiment(exp_args)
+        except BaseException:
```

Comment:
is there no more specific exception you can catch?

Reply:
Since this method calls run_experiment at garage.misc.instrument, I'm not sure about all the possible exceptions that may occur there, and I also want to catch exceptions such as KeyboardInterrupt.
Besides, the purpose of catching this exception is to notify the user of how to restore the working directory in case the restoration performed in the finally blocked is not executed correctly for unexpected reasons, and the exception is raised again after that.